sled-agent/src/profile.rs

```diff
@@ -0,0 +1,273 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Utilities to build SMF profiles.
+
+use illumos_utils::running_zone::InstalledZone;
+use slog::Logger;
+use std::fmt::{Display, Formatter};
+
+pub struct ProfileBuilder {
+    name: String,
+    services: Vec<ServiceBuilder>,
+}
+
+impl ProfileBuilder {
+    pub fn new(name: &str) -> Self {
+        Self { name: name.to_string(), services: vec![] }
+    }
+
+    pub fn add_service(mut self, service: ServiceBuilder) -> Self {
+        self.services.push(service);
+        self
+    }
+
+    pub async fn add_to_zone(
+        &self,
+        log: &Logger,
+        installed_zone: &InstalledZone,
+    ) -> Result<(), std::io::Error> {
+        info!(log, "Profile for {}:\n{}", installed_zone.name(), self);
+
+        let profile_path = format!(
+            "{zone_mountpoint}/{zone}/root/var/svc/profile/site.xml",
+            zone_mountpoint = illumos_utils::zfs::ZONE_ZFS_DATASET_MOUNTPOINT,
+            zone = installed_zone.name(),
+        );
+
+        tokio::fs::write(&profile_path, format!("{self}").as_bytes()).await?;
+        Ok(())
+    }
+}
+
+impl Display for ProfileBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(
+            f,
+            r#"<!DOCTYPE service_bundle SYSTEM "/usr/share/lib/xml/dtd/service_bundle.dtd.1">
+<service_bundle type="profile" name="{name}">
+"#,
+            name = self.name
+        )?;
+        for service in &self.services {
+            write!(f, "{}", service)?;
+        }
+        write!(f, "</service_bundle>")?;
+
+        Ok(())
+    }
+}
+
+pub struct ServiceBuilder {
+    name: String,
+    property_groups: Vec<PropertyGroupBuilder>,
+}
+
+impl ServiceBuilder {
+    pub fn new(name: &str) -> Self {
+        Self { name: name.to_string(), property_groups: vec![] }
+    }
+
+    pub fn add_property_group(
+        mut self,
+        property_group: PropertyGroupBuilder,
+    ) -> Self {
+        self.property_groups.push(property_group);
+        self
+    }
+}
+
+impl Display for ServiceBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(
+            f,
+            r#"  <service version="1" type="service" name="{name}">
+    <instance enabled="true" name="default">
```

Comment:
I think I would be inclined not to assume that all services have only one instance, and that that instance is named **default**.  But this is also not an exported interface so we can totally leave it for now and fix that later as needed.  I might rename the object to `ServiceInstanceBuilder` though maybe.

Reply:
I went ahead and split this out into separate `ServiceBuilder` and `ServiceInstanceBuilder` objects, so that callers can control what instances they end up getting.